crates/turborepo-lib/src/run/watch.rs

```diff
@@ -1,37 +1,231 @@
-            // Should we recover here?
-            let hash = hash.unwrap();
-            let event = hash.event.expect("event is missing");
-            match event {
-                proto::package_change_event::Event::PackageChanged(proto::PackageChanged {
-                    package_name,
-                }) => {
-                    println!("{} changed", package_name);
+        let mut events = client.package_changes().await?;
+        let mut current_runs: HashMap<PackageName, JoinHandle<Result<i32, run::Error>>> =
+            HashMap::new();
+        let event_fut = async {
+            while let Some(event) = events.next().await {
+                let event = event.unwrap();
+                Self::handle_change_event(
+                    &mut run,
+                    event.event.unwrap(),
+                    &mut current_runs,
+                    &base,
+                    &telemetry,
+                    &handler,
+                )
+                .await?;
+            }
+
+            Ok::<(), Error>(())
+        };
+
+        select! {
+            _ = event_fut => {}
+            _ = subscriber.listen() => {
+                tracing::info!("shutting down");
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_change_event(
+        run: &mut Run,
+        event: proto::package_change_event::Event,
+        current_runs: &mut HashMap<PackageName, JoinHandle<Result<i32, run::Error>>>,
+        base: &CommandBase,
+        telemetry: &CommandEventBuilder,
+        handler: &SignalHandler,
+    ) -> Result<(), Error> {
+        // Should we recover here?
+        match event {
+            proto::package_change_event::Event::PackageChanged(proto::PackageChanged {
+                package_name,
+            }) => {
+                let package_name = PackageName::from(package_name);
+                // If not in the filtered pkgs, ignore
+                if !run.filtered_pkgs.contains(&package_name) {
+                    return Ok(());
                 }
-                proto::package_change_event::Event::RediscoverPackages(_) => {
-                    println!("Rediscovering packages");
+
+                let mut args = base.args().clone();
+                args.command = args.command.map(|c| {
+                    if let Command::Watch(execution_args) = c {
+                        Command::Run {
+                            execution_args: Box::new(ExecutionArgs {
+                                filter: vec![format!("...{}", package_name)],
+                                ..*execution_args
+                            }),
+                            run_args: Box::new(RunArgs {
+                                no_cache: true,
+                                daemon: true,
+                                ..Default::default()
+                            }),
+                        }
+                    } else {
+                        unreachable!()
```

Comment:
The lines of code leading up to determining this is a Watch command seems like something we could pull forward so don't need to reparse this out on change events.

Reply:
Ehh kinda. It's a little messy since we need the full args struct from `base`. Either we need to clone the watch part out and keep the args struct around (which makes me a little uncomfortable cause there'd be multiple sources of truth for the watch config) or do this unwrap.